diff --git a/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java b/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java
index 69fbeb7..6282fcb 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/dao/PodcastDao.java
@@ -137,7 +137,7 @@ public class PodcastDao extends AbstractDao {
     public List<PodcastEpisode> getNewestEpisodes(int count) {
         String sql = "select " + EPISODE_QUERY_COLUMNS
                      + " from podcast_episode where status = ? and publish_date is not null " +
-                     "order by publish_date desc limit ?";
+                     "order by publish_date limit ?";
         return query(sql, episodeRowMapper, PodcastStatus.COMPLETED.name(), count);
     }

diff --git a/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java b/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java
index d5fad1d..b155fc4 100644
--- a/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java
+++ b/airsonic-main/src/main/java/org/airsonic/player/service/PodcastService.java
@@ -465,10 +465,10 @@ public class PodcastService {
                 long timeA = a.getPublishDate() == null ? 0L : a.getPublishDate().getTime();
                 long timeB = b.getPublishDate() == null ? 0L : b.getPublishDate().getTime();

-                if (timeA < timeB) {
+                if (timeA > timeB) {
                     return 1;
                 }
-                if (timeA > timeB) {
+                if (timeA < timeB) {
                     return -1;
                 }
                 return 0;
